relay/pkg/node/node.go

```diff
@@ -19,35 +24,54 @@ func Initialize(
 	ctx context.Context,
 	btcChain btc.Handle,
 	hostChain chain.Handle,
-) {
+) *Node {
 	logger.Infof("initializing relay node")
 
-	go runForwarderControlLoop(ctx, btcChain, hostChain)
+	node := &Node{
+		stats: newStats(),
+	}
+
+	go node.runForwarderControlLoop(ctx, btcChain, hostChain)
+
+	return node
 }
 
 // runForwarderControlLoop runs a block forwarder control loop which is
 // responsible for starting the forwarder and acting upon errors by restarting
 // the forwarder instance. The lifecycle of the control loop itself can
 // be managed using the passed context.
-func runForwarderControlLoop(
+func (n *Node) runForwarderControlLoop(
 	ctx context.Context,
 	btcChain btc.Handle,
 	hostChain chain.Handle,
 ) {
 	logger.Infof("running block forwarding")
+	n.stats.notifyBlockForwardingEnabled()
+
+	defer func() {
+		logger.Infof("stopping block forwarding")
+		n.stats.notifyBlockForwardingDisabled()
+	}()
 
 	for {
-		forwarder := block.RunForwarder(ctx, btcChain, hostChain)
+		forwarder := block.RunForwarder(ctx, btcChain, hostChain, n.stats)
 
 		select {
 		case err := <-forwarder.ErrChan():
 			logger.Errorf(
 				"error occurred during block forwarding: [%v]",
 				err,
 			)
+
+			n.stats.notifyBlockForwardingErrored()
 		case <-ctx.Done():
 			logger.Infof("stopping block forwarding")
 			return
 		}
 	}
 }
+
+// Stats returns relay node statistics.
+func (n *Node) Stats() *Stats {
```

Comment:
I'd consider returning an interface instead of the struct. We'd not have to worry about `NotifyBlockPulled` and `NotifyBlocksPushed` being exported and that they can be called by the code that shouldn't do it. Then, we can pass around the structure to `RunForwarder` without having to define an interface there and mocking it in tests.

Reply:
I agree that `NotifyBlockPulled` and `NotifyBlocksPushed` should be protected. I think we can introduce some kind of `node.ReadonlyStats` interface and return it from `Stats` method. However, regarding `RunForwarding` I'm strongly in favor of leaving as it is. When `block.RunForwarder` accepts `block.ForwarderObserver` we are not tied to any implementation, if we need to pass another kind of observer or register multiple ones, it's easy to do. This approach is also idiomatic in Go. If we want to pass `node.Stats` struct to `RunForwarder` we need to move that struct somewhere out of the `node` pkg to avoid the import cycle. Also, we will tie the forwarder with `node.Stats` implementation. This is a quite high price to get rid of the test mock which is really not a problem here.